qemu_v8: Add support of xen and virtualization

ruchi393 opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

ruchi393 commented:
This PR has a related PR in manifest:
https://github.com/OP-TEE/manifest/pull/186

This PR is dependent on another PR where the first 2 patches for uboot and GICv3 are in review:
https://github.com/OP-TEE/build/pull/492

jforissier commented on qemu_v8.mk:
```diff
@@ -261,6 +286,9 @@ linux-cleaner: linux-cleaner-common
 ################################################################################
 OPTEE_OS_COMMON_FLAGS += DEBUG=$(DEBUG) CFG_ARM_GICV3=$(GICV3)
+ifeq ($(XEN_BOOT),y)
+OPTEE_OS_COMMON_FLAGS += CFG_VIRTUALIZATION=y CFG_CORE_ASLR=y CFG_TA_ASLR=y CFG_CORE_DYN_SHM=y
```
I think CFG_CORE_ASLR=y CFG_TA_ASLR=y should be removed, because (1) they are the default values for QEMUv8 and (2) it should be possible to disable them if needed (they are not mandatory for virtualization, are they?)

ruchi393 replied:
No they are not. I added them I was playing with different options. I will remove them.

jforissier commented on qemu_v8.mk:
```diff
@@ -313,6 +379,17 @@ run: all
+QEMU_MEM ?= 1057
+
+ifeq ($(XEN_BOOT),y)
+QEMU_MEM 	= 2048
+QEMU_SMP	= 4
```
?=

ruchi393 replied:
ok

jforissier commented on qemu_v8.mk:
```diff
@@ -313,6 +379,17 @@ run: all
+		   -machine virt,gic-version=$(QEMU_GIC_VERSION) \
+		   -drive if=none,file=$(XEN_EXT4),format=raw,id=hd1 \
+		   -device virtio-blk-device,drive=hd1
+endif
```
else
QEMU_SMP ?= 2
QEMU_MEM ?= 1057

ruchi393 replied:
ok

jenswi-linaro replied:
It would be nice to get rid of this `1057`, but perhaps in a separate PR.

jforissier reviewed (commented):
About commit "qemu_v8.mk: Enable xen tools in DOM0 rootfs":

> xen tools are required to launch the other DOMU's from

Xen, DomU

> DOM0. Enable then in buildroot. Since XEN is built outside

Dom0, Buildroot

> buildroot, it is essential to maintain the same version of
> xen built in build system and xen-tools built in buildroot.
>
> DOMU needs a rootfs without the xen tools. So another target
> is introduced to build rootfs for DOMU's without xen tools.

Does this commit allow to run `xtest` in DomU? I understand that `make run` launches Dom0. What would be the additional steps to launch one or more DomU's?

How to run `make check` in Dom0 or DomU? (I am very interesting by this for release testing)

ruchi393 commented:
> About commit "qemu_v8.mk: Enable xen tools in DOM0 rootfs":
> 
> > xen tools are required to launch the other DOMU's from
> 
> Xen, DomU
> 
> > DOM0. Enable then in buildroot. Since XEN is built outside
> 
> Dom0, Buildroot
> 
> > buildroot, it is essential to maintain the same version of
> > xen built in build system and xen-tools built in buildroot.
> > DOMU needs a rootfs without the xen tools. So another target
> > is introduced to build rootfs for DOMU's without xen tools.
> 
> Does this commit allow to run `xtest` in DomU? I understand that `make run` launches Dom0. What would be the additional steps to launch one or more DomU's?
Yes xtest run finw tih this. Test 1013.2 fails due to lack of secure memory.
I have tried a single DomU for now. To do that, the steps are
You need to have QEMU_VIRTFS enabled.
On Dom0
mkdir -p /host && mount -t 9p -o trans=virtio host /host
cd /host/build/qemu_v8/xen
xl create guest.cfg

To got to console of this domU
xl console domU

I am also toying around with the idea of putting the domU cfg and files in an ext2 partition which we can then mount and run from Dom0.

Eventually it would be good if we can add the flow in our CI's.

> 
> How to run `make check` in Dom0 or DomU? (I am very interesting by this for release testing)
I will need to explore that.  I haven;t looked at what  `make check` does. Will have a look and see what I can add over there.

ruchi393 commented on qemu_v8.mk:
```diff
@@ -303,6 +331,44 @@ urootfs: buildroot $(BINARIES_PATH)
+# When creating the image containing Linux kernel, we need to temporarily store
+# the files somewhere.
+$(XEN_TMP):
+	mkdir -p $@
```
ok

jforissier replied:
Another option would be to use an order-only prerequisite:
```
$(XEN_TMP):
	mkdir $@

xen-create-image: xen linux buildroot | $(XEN_TMP)
	...
```
See https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html.
That's a nice trick I think, worth remembering ;) that being said no reason to introduce this pattern here since we're using unconditional `mkdir -p`'s everywhere in the OP-TEE projects.

ruchi393 replied:
Thanks @jforissier. This is indeed nice trick to remember

ruchi393 commented on qemu_v8.mk:
```diff
@@ -303,6 +331,44 @@ urootfs: buildroot $(BINARIES_PATH)
+xen-create-image: $(XEN_TMP) xen linux buildroot
+	# Use a written path to avoid rm -f real host machine files (in case
+	# GRUB2_TMP has been set to an empty string)
+	rm -f $(BINARIES_PATH)/xen_files/*
```
Right. I will remove this and directly add mkdir here.

jforissier commented:
> > About commit "qemu_v8.mk: Enable xen tools in DOM0 rootfs":
> > > xen tools are required to launch the other DOMU's from
> > 
> > 
> > Xen, DomU
> > > DOM0. Enable then in buildroot. Since XEN is built outside
> > 
> > 
> > Dom0, Buildroot
> > > buildroot, it is essential to maintain the same version of
> > > xen built in build system and xen-tools built in buildroot.
> > > DOMU needs a rootfs without the xen tools. So another target
> > > is introduced to build rootfs for DOMU's without xen tools.
> > 
> > 
> > Does this commit allow to run `xtest` in DomU? I understand that `make run` launches Dom0. What would be the additional steps to launch one or more DomU's?
> > Yes xtest run finw tih this. Test 1013.2 fails due to lack of secure memory.
> > I have tried a single DomU for now. To do that, the steps are
> > You need to have QEMU_VIRTFS enabled.
> > On Dom0
> > mkdir -p /host && mount -t 9p -o trans=virtio host /host

Note, with `QEMU_VIRTFS_AUTOMOUNT=y` the host folder is automatically mounted under `/mnt/host` in QEMU ;)

> > cd /host/build/qemu_v8/xen
> > xl create guest.cfg
> 
> To got to console of this domU
> xl console domU

That is good! Perhaps some documentation update would help (https://optee.readthedocs.io/en/latest/architecture/virtualization.html), but as a first step I think your explanation above would be great to have in the commit description.

> 
> I am also toying around with the idea of putting the domU cfg and files in an ext2 partition which we can then mount and run from Dom0.

Would it simplify things compared to the 9p mount?

> Eventually it would be good if we can add the flow in our CI's.

Sure!
 
> > How to run `make check` in Dom0 or DomU? (I am very interesting by this for release testing)
> > I will need to explore that.  I haven;t looked at what  `make check` does. Will have a look and see what I can add over there.

If we could make it run virtualized with a simple command line argument that would be great. FYI, here is how I am doing release testing with QEMU/QEMUv8. Could be automated further but you get the idea.

```
$ cat configs.txt

CFG_WITH_PAGER=y
CFG_RPMB_FS=y CFG_RPMB_WRITE_KEY=y
CFG_REE_FS=n CFG_RPMB_FS=y CFG_RPMB_WRITE_KEY=y
CFG_CORE_DYN_SHM=n
CFG_WITH_PAGER=y CFG_RPMB_FS=y CFG_RPMB_WRITE_KEY=y
CFG_ULIBS_SHARED=y
CFG_ULIBS_SHARED=y CFG_WITH_PAGER=y
CFG_FTRACE_SUPPORT=y CFLAGS_ta_arm32=-pg CFLAGS_ta_arm64=-pg
COMPILER=clang
$ (cat configs.txt | while read c; do rm -rf ../out-br/build/optee* ../optee_os/out; echo "+++++ [$c]"; make -j10 $c check; done;) | tee xtest.log
...
$ grep -E '(^Status:|^\+\+\+\+\+| FAIL|!!! )' xtest.log
+++++ []
Status: PASS (3246 test cases)
+++++ [CFG_WITH_PAGER=y]
Status: PASS (3221 test cases)
+++++ [CFG_RPMB_FS=y CFG_RPMB_WRITE_KEY=y]
Status: PASS (3279 test cases)
+++++ [CFG_REE_FS=n CFG_RPMB_FS=y CFG_RPMB_WRITE_KEY=y]
Status: PASS (3246 test cases)
...
```

ruchi393 commented:
All review comments addressed, make check option modified for XEN_BOOT
Tests 1013.2 and 1034 need to be disabled due to lack of secure memory for guests. I will be sending out another patch for the same.

This PR also has uboot patch in it as it is a dependency. Will remove it once that PR gets merged.

jenswi-linaro commented on kconfigs/xen.conf:
```diff
@@ -0,0 +1,30 @@
+CONFIG_TEE=y
+CONFIG_OPTEE=y
+
+# CONFIG_ACPI is not set
```
Do we need all these commented out config variables? We don't keep them for the kernel or U-Boot.

ruchi393 replied:
These are needed here as when we enable CONFIG_EXPERT option (required for TEE and OPTEE) , 
these configs get added and default values of these options is not assumed and build stops to ask user for these.
I didn't explore it much but at that time adding these options here stopped build from halting to ask for these values. I will explore xen a little more to see if we can do it in a better way.

ruchi393 replied:
I checked in xen again , few of the config options like ACPI and others mentioned in the conf file don't have default values set and are enabled only with EXPERT option. So, we will have to add them in .conf file else build will stop to ask for these options.

jenswi-linaro replied:
That's fine, my point was to have just the ones needed or this config might only work with the the current Xen version, which would make upgrades harder than necessary.

ruchi393 replied:
If I keep only the EXPERT, TEE and OPTEE options, xen build will stop in middle to ask for these config values  which user will have to  select. Would that work with the CI ?

jenswi-linaro replied:
No, it must be fully automated. Worst case you need to keep this, but then you'll need to leave detailed instructions on how to generate that this file so it can be re-generated when upgrading Xen. Surely there must be a nice way of doing this.

ruchi393 replied:
I double checked win xen repo. Right now TEE Mediator support in XEN is in Tech Preview stage. Hence it can be enabled only if CONFIG_EXPERT=y is selected i.e expert mode. With expert mode, this issue will remain. As more features are added in subsequent xen releases for expert mode, user will be prompted for the options which don't have default values in xen kconfig files and either these options have to be manually selected or corresponding option has to be added in the build/kconfig/xen.conf for now.

jforissier commented on qemu_v8.mk:
```diff
@@ -189,6 +298,68 @@ soc-term:
+XEN_TMP ?= $(BINARIES_PATH)/xen_files
+
+$(XEN_TMP):
+	mkdir -p @
```
`$@`

ruchi393 replied:
Sorry, I had this change done locally but forgot to add it in the patch. Will correct it.

jenswi-linaro commented on qemu-check.exp:
```diff
@@ -45,7 +45,11 @@ open "serial1.log" "w+"
 if {[string first "aarch64" $::env(QEMU)] != -1} {
-	spawn $::env(QEMU) -nographic -serial mon:stdio -serial file:serial1.log -smp $::env(QEMU_SMP) -machine virt,secure=on -cpu cortex-a57 -machine virt,gic-version=$::env(QEMU_GIC) -d unimp -semihosting-config enable=on,target=native -m 1057 -bios bl1.bin -initrd rootfs.cpio.gz -kernel Image -no-acpi -append "console=ttyAMA0,38400 keep_bootcon root=/dev/vda2"
+	if {$::env(XEN_BOOT) == "y"} {
+		spawn $::env(QEMU) -nographic -serial mon:stdio -serial file:serial1.log -smp $::env(QEMU_SMP) -machine virt,secure=on -cpu cortex-a57 -machine virt,gic-version=$::env(QEMU_GIC) -d unimp -semihosting-config enable=on,target=native -m $::env(QEMU_MEM) -bios bl1.bin -initrd rootfs.cpio.gz -kernel Image -no-acpi -append "console=ttyAMA0,38400 keep_bootcon root=/dev/vda2" -machine virtualization=tr ...
```
Would it make sense to move `-machine virtualization=true` into the first `-machine` option instead?

ruchi393 replied:
I have changed this.

jforissier reviewed (commented):
- The doc will need updating:
```
/bin/bash: line 1: virt-make-fs: command not found
make: *** [Makefile:358: xen-create-image] Error 127
make: *** Waiting for unfinished jobs....
$ virt-make-fs
Command 'virt-make-fs' not found, but can be installed with:
sudo apt install libguestfs-tools
```
- I hit the following error:
```
$ make -j10 xen-create-image XEN_BOOT=y
...
libguestfs: error: /usr/bin/supermin exited with error status 1.
To see full error messages you may need to enable debugging.
Do:
  export LIBGUESTFS_DEBUG=1 LIBGUESTFS_TRACE=1
and run the command again.  For further information, read:
  http://libguestfs.org/guestfs-faq.1.html#debugging-libguestfs
You can also run 'libguestfs-test-tool' and post the *complete* output
into a bug report or message to the libguestfs mailing list.
make: *** [Makefile:358: xen-create-image] Error 1


$ export LIBGUESTFS_DEBUG=1 LIBGUESTFS_TRACE=1
$ make -j10 xen-create-image XEN_BOOT=y
...
supermin: kernel: looking for kernel using environment variables ...
supermin: kernel: looking for kernels in /lib/modules/*/vmlinuz ...
supermin: kernel: looking for kernels in /boot ...
supermin: kernel: kernel version of /boot/vmlinuz-5.11.0-18-generic = 5.11.0-18-generic (from filename)
supermin: kernel: picked modules path /lib/modules/5.11.0-18-generic
supermin: kernel: kernel version of /boot/vmlinuz-5.8.0-55-generic = 5.8.0-55-generic (from filename)
supermin: kernel: picked modules path /lib/modules/5.8.0-55-generic
supermin: kernel: picked vmlinuz /boot/vmlinuz-5.11.0-18-generic
supermin: kernel: kernel_version 5.11.0-18-generic
supermin: kernel: modpath /lib/modules/5.11.0-18-generic
cp: cannot open '/boot/vmlinuz-5.11.0-18-generic' for reading: Permission denied
supermin: cp -p '/boot/vmlinuz-5.11.0-18-generic' '/var/tmp/.guestfs-1000/appliance.d.5x7g14be/kernel': command failed, see earlier errors
libguestfs: error: /usr/bin/supermin exited with error status 1, see debug messages above
libguestfs: trace: launch = -1 (error)
libguestfs: trace: close
libguestfs: closing guestfs handle 0x556bdd45cf60 (state 0)
libguestfs: command: run: rm
libguestfs: command: run: \ -rf /tmp/libguestfsLHACLu
make: *** [Makefile:358: xen-create-image] Error 1
```

jenswi-linaro commented on qemu_v8.mk:
```diff
@@ -19,6 +19,15 @@ BR2_ROOTFS_POST_SCRIPT_ARGS = "$(QEMU_VIRTFS_AUTOMOUNT) $(QEMU_VIRTFS_MOUNTPOINT
 OPTEE_OS_PLATFORM = vexpress-qemu_armv8a
 
+################################################################################
+# If you change this, you MUST run `make clean` before rebuilding
```
It might be worth mentioning which gits must be cleaned. Does the kernel need cleaning? Does buildroot need cleaning?

ruchi393 replied:
Sure, I will add that. gits which need cleaning are mainly - optee_os, arm-tf, buildroot

ruchi393 replied:
buildroot clean is not required. The xen scripts don't interfere with normal build. so, only arm-tf-clean and optee-os-clean needs to be done. I will update the comment.

jenswi-linaro commented on qemu_v8.mk:
```diff
@@ -255,6 +280,9 @@ linux-cleaner: linux-cleaner-common
 ################################################################################
 OPTEE_OS_COMMON_FLAGS += DEBUG=$(DEBUG) CFG_ARM_GICV3=$(GICV3)
+ifeq ($(XEN_BOOT),y)
+OPTEE_OS_COMMON_FLAGS += CFG_VIRTUALIZATION=y CFG_CORE_DYN_SHM=y
```
Isn't `CFG_CORE_DYN_SHM=y` the default value already?

ruchi393 replied:
Yes, I will remove it.

jenswi-linaro reviewed (commented):
Comments for the commit "qemu_v8.mk: Add option to add XEN in boot flow"

jforissier commented on qemu_v8.mk:
```diff
@@ -297,6 +325,37 @@ $(ROOTFS_UGZ): u-boot buildroot | $(BINARIES_PATH)
+	virt-make-fs -t vfat $(XEN_TMP) $(XEN_EXT4)
+
+xen-distclean:
+	$(MAKE) -C $(XEN_PATH) distclean
```
Why `distclean`? Isn't `clean` good enough?

ruchi393 replied:
clean should do the work. Will change.

jenswi-linaro reviewed (commented):
Comments for the commit "qemu_v8.mk: Add option to add XEN in boot flow"

In the commit message:
`s/Enable then/Enable them/

jforissier commented on qemu_v8.mk:
```diff
@@ -297,6 +325,37 @@ $(ROOTFS_UGZ): u-boot buildroot | $(BINARIES_PATH)
+	XEN_TARGET_ARCH=arm64 \
+	CONFIG_XEN_INSTALL_SUFFIX=.gz	\
+	CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)"
+	ln -sf $(XEN_PATH)/xen $(BINARIES_PATH)
```
That is causing some rebuild of the `.config` every time `make xen-common` is run and seems overly complicated.
How about:
```make
$(XEN_PATH)/xen/.config:
        $(MAKE) -C $(XEN_PATH)/xen XEN_TARGET_ARCH=arm64 defconfig
        cd $(XEN_PATH)/xen && \
        tools/kconfig/merge_config.sh -m .config $(ROOT)/build/kconfigs/xen.conf

xen: $(XEN_PATH)/xen/.config | $(BINARIES_PATH)
        $(MAKE) -C $(XEN_PATH) dist-xen \
        XEN_TARGET_ARCH=arm64 \
        CONFIG_XEN_INSTALL_SUFFIX=.gz   \
        CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)"
        ln -sf $(XEN_PATH)/xen $(BINARIES_PATH)
```

ruchi393 replied:
Will change

jforissier commented on qemu-check.exp:
```diff
@@ -72,47 +121,36 @@ send -- "export LD_LIBRARY_PATH=/lib:/lib/arm-linux-gnueabihf\r"
-			set star 1
+check_test_result $teecore
+if {$::env(XEN_BOOT) == "y"} {
+	info " Booting DomU.\n"
```
Could we have a message to tell we're running Dom0 in the first part of the test above? Something like:
```
info " done, guest is booted"
if {$::env(XEN_BOOT) == "y"} {
        info " (Xen Dom0)"
}
info ".\n"
```

ruchi393 replied:
Will change

ruchi393 commented on qemu-check.exp:
```diff
@@ -45,7 +45,11 @@ open "serial1.log" "w+"
+		spawn $::env(QEMU) -nographic -serial mon:stdio -serial file:serial1.log -smp $::env(QEMU_SMP) -machine virt,secure=on -cpu cortex-a57 -machine virt,gic-version=$::env(QEMU_GIC) -d unimp -semihosting-config enable=on,target=native -m $::env(QEMU_MEM) -bios bl1.bin -initrd rootfs.cpio.gz -kernel Image -no-acpi -append "console=ttyAMA0,38400 keep_bootcon root=/dev/vda2" -machine virtualization=tr ...
+	} else {
+		spawn $::env(QEMU) -nographic -serial mon:stdio -serial file:serial1.log -smp $::env(QEMU_SMP) -machine virt,secure=on -cpu cortex-a57 -machine virt,gic-version=$::env(QEMU_GIC) -d unimp -semihosting-config enable=on,target=native -m $::env(QEMU_MEM) -bios bl1.bin -initrd rootfs.cpio.gz -kernel Image -no-acpi -append "console=ttyAMA0,38400 keep_bootcon root=/dev/vda2"
+	}
```
Sure, I will change it.

ruchi393 replied:
@jforissier , I tried this change. However I end up with a spawn error.

`if {[string first "aarch64" $::env(QEMU)] != -1} {
        set xen_boot_args ""
        if {$::env(XEN_BOOT) == "y"} {
                set xen_boot_args "-machine virtualization=true -drive if=none,file=xen.ext4,format=raw,id=hd1 -device virtio-blk-device,drive=hd1 -fsdev local,id=fsdev0,path=../..,security_model=none -device virtio-9p-device,fsdev=fsdev0,mount_tag=host"
        }
        spawn $::env(QEMU) -nographic -serial mon:stdio -serial file:serial1.log -smp $::env(QEMU_SMP) -machine virt,secure=on -cpu cortex-a57 -machine virt,gic-version=$::env(QEMU_GIC) -d unimp -semihosting-config enable=on,target=native -m $::env(QEMU_MEM) -bios bl1.bin -initrd rootfs.cpio.gz -kernel Image -no-acpi $xen_boot_args -append "console=ttyAMA0,38400 keep_bootcon root=/dev/vda2"
} else {
        spawn $::env(QEMU) -nographic -monitor none -machine virt -machine secure=on -cpu cortex-a15 -smp $::env(QEMU_SMP) -d unimp -semihosting-config enable=on,target=native -m 1057 -serial stdio -serial file:serial1.log -bios $bios
}`

Error in serial0.log
spawn [open ...]^M
spawn /home/ruchika/OPTEE-XEN-FINAL/build/../qemu/build/aarch64-softmmu/qemu-system-aarch64 -nographic -serial mon:stdio -serial file:serial1.log -smp 4 -machine virt,secure=on -cpu cortex-a57 -machine virt,gic-version=3 -d unimp -semihosting-config enable=on,target=native -m 2048 -bios bl1.bin -initrd rootfs.cpio.gz -kernel Image -no-acpi -machine virtualization=true -drive if=none,file=xen.ext4,format=raw,id=hd1 -device virtio-blk-device,drive=hd1 -fsdev local,id=fsdev0,path=../..,security_model=none -device virtio-9p-device,fsdev=fsdev0,mount_tag=host -append console=ttyAMA0,38400 keep_bootcon root=/dev/vda2^M
qemu-system-aarch64: -machine virtualization=true -drive if=none,file=xen.ext4,format=raw,id=hd1 -device virtio-blk-device,drive=hd1 -fsdev local,id=fsdev0,path=../..,security_model=none -device virtio-9p-device,fsdev=fsdev0,mount_tag=host: invalid option^M
send: spawn id exp9 not open
    while executing
"send -- "root\r\r""
    (file "/home/ruchika/OPTEE-XEN-FINAL/build/../build/qemu-check.exp" line 116)^M

jforissier replied:
OK it must be some kind of Tcl syntax issue... Never mind.

ruchi393 commented:
> * The doc will need updating:
> 
> ```
> /bin/bash: line 1: virt-make-fs: command not found
> make: *** [Makefile:358: xen-create-image] Error 127
> make: *** Waiting for unfinished jobs....
> $ virt-make-fs
> Command 'virt-make-fs' not found, but can be installed with:
> sudo apt install libguestfs-tools
> ```
> 
> * I hit the following error:
> 
> ```
> $ make -j10 xen-create-image XEN_BOOT=y
> ...
> libguestfs: error: /usr/bin/supermin exited with error status 1.
> To see full error messages you may need to enable debugging.
> Do:
>   export LIBGUESTFS_DEBUG=1 LIBGUESTFS_TRACE=1
> and run the command again.  For further information, read:
>   http://libguestfs.org/guestfs-faq.1.html#debugging-libguestfs
> You can also run 'libguestfs-test-tool' and post the *complete* output
> into a bug report or message to the libguestfs mailing list.
> make: *** [Makefile:358: xen-create-image] Error 1
> 
> 
> $ export LIBGUESTFS_DEBUG=1 LIBGUESTFS_TRACE=1
> $ make -j10 xen-create-image XEN_BOOT=y
> ...
> supermin: kernel: looking for kernel using environment variables ...
> supermin: kernel: looking for kernels in /lib/modules/*/vmlinuz ...
> supermin: kernel: looking for kernels in /boot ...
> supermin: kernel: kernel version of /boot/vmlinuz-5.11.0-18-generic = 5.11.0-18-generic (from filename)
> supermin: kernel: picked modules path /lib/modules/5.11.0-18-generic
> supermin: kernel: kernel version of /boot/vmlinuz-5.8.0-55-generic = 5.8.0-55-generic (from filename)
> supermin: kernel: picked modules path /lib/modules/5.8.0-55-generic
> supermin: kernel: picked vmlinuz /boot/vmlinuz-5.11.0-18-generic
> supermin: kernel: kernel_version 5.11.0-18-generic
> supermin: kernel: modpath /lib/modules/5.11.0-18-generic
> cp: cannot open '/boot/vmlinuz-5.11.0-18-generic' for reading: Permission denied
> supermin: cp -p '/boot/vmlinuz-5.11.0-18-generic' '/var/tmp/.guestfs-1000/appliance.d.5x7g14be/kernel': command failed, see earlier errors
> libguestfs: error: /usr/bin/supermin exited with error status 1, see debug messages above
> libguestfs: trace: launch = -1 (error)
> libguestfs: trace: close
> libguestfs: closing guestfs handle 0x556bdd45cf60 (state 0)
> libguestfs: command: run: rm
> libguestfs: command: run: \ -rf /tmp/libguestfsLHACLu
> make: *** [Makefile:358: xen-create-image] Error 1
> ```
Thanks, I did get this problem initially but got fixed after doing this as suggested by @jbech-linaro.

The reason for that is because your /boot/vmlinuz-* files are only readable by the root user. To work around this, you need to make them readable. Note that after upgrading to a new kernel on your host, you'll have to redo this again (and again).

$ sudo chmod 644 /boot/vmlinuz-`uname -r`

Should I add this information in the "README.md" or in optee-docs ?

jforissier commented:
> The reason for that is because your /boot/vmlinuz-* files are only readable by the root user.

Why should the build of the Xen VM be given access to anything under my `/boot` directory?

jforissier commented:
Some comments about the last commit description:
```
 qemu_v8.mk: Enable xen tools in Dom0 rootfs
```
qemu_v8: xen: Enable DomU

```
xen tools are required to launch the other DomU's from
Dom0. Enable then in Buildroot. Since XEN is built outside
```
Xen tools, Xen
```
buildroot, it is essential to maintain the same version of
```
Buildroot
```
xen built in build system and xen-tools built in buildroot.
```
Xen, Xen tools, Buildroot
```
DomU needs a rootfs without the xen tools. So another target
```
Xen. Would it really be a problem to use the same rootfs? Not a real world scenario of course, and the Xen tools would indeed be useless, but it could save some significant build time and space perhaps?

```
is introduced to build rootfs for DOMU's without xen tools.
```
DomU
```
To launch & run DomU:

make XEN_BOOT=y run

From Dom0 prompt:
```
From a Dom0 root prompt:
```
> mkdir -p /host && mount -t 9p -o trans=virtio host /host
```
Not needed, host directory is automatically mounted as /mnt/host

```
> cd /host/qemu_v8/xen
```
cd /mnt/host/build/qemu_v8/xen
```
> xl create guest.cfg	# This will create a DomU
To goto DomU's console:
> xl console domu

To run and check xtests from Dom0 and DomU:
make XEN_BOOT=y check
```

@ruchi393 sorry for throwing so many comments at you, I added them as I am testing the PR on my laptop. But thanks for putting this together, that is a pretty cool test environment we have now :wink: and I will feel much better being able to test virtualization from time to time.

ruchi393 commented:
> make XEN_BOOT=y check
> ```
> 
> @ruchi393 sorry for throwing so many comments at you, I added them as I am testing the PR on my laptop. But thanks for putting this together, that is a pretty cool test environment we have now  and I will feel much better being able to test virtualization from time to time.

@jforissier , no issues at all. Thanks a lot for taking out time to review and test it. I will make the changes and re-submit.

ruchi393 commented:
> Xen. Would it really be a problem to use the same rootfs? Not a real world scenario of course, and the Xen tools would indeed be useless, but it could save some significant build time and space perhaps?

Certain xen scripts get added in init.d which are supposed to be run only by Dom0 . This causes a kernel crash when trying to run /init scripts from DomU. As a result I had to add a buildroot target which doesn't build in xen tools

ruchi393 commented:
Thanks @jforissier and @jenswi-linaro for helping in review.
I have modified the commit message as per suggestions. I have also addressed most of the comments. Will squash these changes after review.

jforissier commented:
> > Xen. Would it really be a problem to use the same rootfs? Not a real world scenario of course, and the Xen tools would indeed be useless, but it could save some significant build time and space perhaps?
> 
> Certain xen scripts get added in init.d which are supposed to be run only by Dom0 . This causes a kernel crash when trying to run /init scripts from DomU. As a result I had to add a buildroot target which doesn't build in xen tools

A kernel crash isn't a good response to an unsupported action, but that's another story ;-) let's keep things like that then.

jenswi-linaro commented on common.mk:
```diff
@@ -323,6 +332,31 @@ buildroot-clean:
 buildroot-cleaner:
 	@rm -rf $(ROOT)/out-br
 
+.PHONY: buildroot_domu
```
buildroot-domu

ruchi393 replied:
Fixed

jenswi-linaro commented on common.mk:
```diff
@@ -323,6 +332,31 @@ buildroot-clean:
+		--make-cmd $(MAKE))
+	@$(MAKE) -C ../out-br-domu all
+
+.PHONY: buildroot-clean
```
buildroot-domu-clean

ruchi393 replied:
Fixed

jenswi-linaro commented on common.mk:
```diff
@@ -323,6 +332,31 @@ buildroot-clean:
+buildroot-domu-clean:
+	@test ! -d $(ROOT)/out-br-domu || $(MAKE) -C $(ROOT)/out-br-domu clean
+
+.PHONY: buildroot-cleaner
```
buildroot-domu-cleaner

jforissier commented on qemu_v8.mk:
```diff
@@ -285,6 +327,43 @@ $(ROOTFS_UGZ): u-boot buildroot | $(BINARIES_PATH)
+	XEN_TARGET_ARCH=arm64 \
+	CONFIG_XEN_INSTALL_SUFFIX=.gz	\
+	CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)"
+	ln -sf $(XEN_IMAGE) $(BINARIES_PATH)
```
Is this symlink needed? It looks like `make XEN_BOOT=y run-only` works fine without `xen.efi` in `$(BINARIES_PATH)`.

ruchi393 replied:
This was mainly required when I was using qemu options to load xen for edk2 boot. I will remove it for now and add it later when I add support to boot xen via edk2.

jforissier replied:
Oh I see, one more configuration to test :open_mouth: :wink: 
Yes better to remove it if not used for now.

jforissier commented on qemu-check.exp:
```diff
@@ -45,7 +94,11 @@ open "serial1.log" "w+"
+	if {$::env(XEN_BOOT) == "y"} {
+		spawn $::env(QEMU) -nographic -serial mon:stdio -serial file:serial1.log -smp $::env(QEMU_SMP) -machine virt,secure=on,gic-version=$::env(QEMU_GIC),virtualization=true -cpu cortex-a57 -d unimp -semihosting-config enable=on,target=native -m $::env(QEMU_MEM) -bios bl1.bin -initrd rootfs.cpio.gz -kernel Image -no-acpi -append "console=ttyAMA0,38400 keep_bootcon root=/dev/vda2" -drive if=none,file= ...
+	} else {
+		spawn $::env(QEMU) -nographic -serial mon:stdio -serial file:serial1.log -smp $::env(QEMU_SMP) -machine virt,secure=on,virt,gic-version=$::env(QEMU_GIC) -cpu cortex-a57 -d unimp -semihosting-config enable=on,target=native -m $::env(QEMU_MEM) -bios bl1.bin -initrd rootfs.cpio.gz -kernel Image -no-acpi -append "console=ttyAMA0,38400 keep_bootcon root=/dev/vda2"
```
You have introduced `virt` twice here: `-machine virt,secure=on,virt,gic-version=$::env(QEMU_GIC)`, that doesn't work.

ruchi393 replied:
Sorry, I will correct it. Must have happened during rebasing.

jforissier commented on qemu_v8/xen/guest.cfg:
```diff
@@ -0,0 +1,8 @@
+ramdisk = "/mnt/host/out-br-domu/images/rootfs.cpio.gz"
+name="domu"
+tee="optee"
+
```
Extra blank line

jforissier reviewed (approved):
"qemu_v8.mk: Add option to add XEN in boot flow" -> s/XEN/Xen/
Plus one minor comment below. In any case:

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

ruchi393 commented:
Thanks @jforissier .  Commit message fixed ,blank line removed and tag applied.

jenswi-linaro commented:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

ruchi393 commented:
Tags applied.
